### PR TITLE
Human tracking multiview triangulation

### PR DESCRIPTION
**Pull request by ptc-mklaudiny**
I introduced new methods of fusing human objects from multiple views. More testing and tuning is needed to select the most reliable one in near term (eg. LiveWorx). It's easy to switch between them through flags without changing the current behaviour. I would like to have this merged since lot of the functionality will be useful down the line. 

Also, added new dependency  ml-matrix for linear algebra.

### CODE REVIEW

File: libraries/HumanPoseFuser.js [hunk 1]
```diff
@@ -403,38 +550,51 @@ class HumanPoseFuser {
...
+     * @param {number} distance - distance between two poses (as computed by calculateDistanceBetweenPoses)
+     * @param {number} ts1 - timestamp of first pose
+     * @param {number} ts2 - timestamp of second pose
+     * @param {number} multiplier - multiplier of distance threshold
+     * @return {boolean}
+     */
+    arePosesOfSamePerson(distance, ts1, ts2, multiplier = 1.0) {
+        // distance threshold is calculated from two components:
+        // - base deviation between poses at the same timestamp
+        // - max possible shift between poses from different timestamp due to person's movement
```
Comment (hobinjk-ptc) [inline libraries/HumanPoseFuser.js:563]: nit: trailing spaces (see Action output below)

File: libraries/HumanPoseFuser.js [hunk 2]
```diff
@@ -526,16 +686,242 @@ class HumanPoseFuser {
...
+    triangulatePose(poseDataArr) {
+        if (poseDataArr.length < 2) {
+            return null;
+        }
+
+        // prepare data
+        let projectionMatrices = []; // per-view
+        let jointsInView = []; // joint 2D positions + confidences per view
+        let latestTS = 0; // the latest data timestamp
+        for (let [id, pose] of poseDataArr) {
```
Comment (hobinjk-ptc) [inline libraries/HumanPoseFuser.js:767]: another nit inline, could be _id to make it clear that you don't need the id from poseDataArr's entries

File: libraries/HumanPoseFuser.js [hunk 3]
```diff
@@ -568,33 +1015,29 @@ class HumanPoseFuser {
...
-        this.objectsRef[fusedObjectId].updateJoints(poseData[selectedObjectId].joints);
-        this.objectsRef[fusedObjectId].lastUpdateDataTS = poseData[selectedObjectId].timestamp;
+        this.objectsRef[fusedObjectId].updateJoints(finalPose.joints);
+        this.objectsRef[fusedObjectId].lastUpdateDataTS = finalPose.timestamp;
         server.resetObjectTimeout(fusedObjectId); // keep the object alive
 
-        // create copy and update name (confidence attribute not copied)
-        let wholePose = Object.assign({}, poseData[selectedObjectId]);
+        // create copy and update name (unnecessary properties not copied)
+        let wholePose = Object.assign({}, finalPose);
```
Comment (hobinjk-ptc) [inline libraries/HumanPoseFuser.js:1035]: I don't see any bugs resulting from this but note that Object.assign is a shallow copy, for example mutating finalPose.joints will change wholePose.joints and vice versa

**Review by hobinjk-ptc — APPROVED** [review]
Good work! Tested this out locally with 2 devices and LGTM

File: libraries/HumanPoseFuser.js (hunk 3 above)
Comment (ptc-mklaudiny) [reply libraries/HumanPoseFuser.js:1035]: It should be fine, the pose data like joints are not expected to change one generated. In some cases, internal 'joints' array might be pointed at from several WholePoseData objects with different lifetimes. But I assume that garbage collector does a right thing and removes it once there is no reference to it.

**Pull request merged**